### PR TITLE
fix(sql): bind variable creation wasn't using correct execution context

### LINKED ISSUE #1740: NullPointerException using PostgreSQL/JDBC PreparedStatement on table with 'symbol index'
### Describe the bug

When using PreparedStatements with JDBC on a table with `symbol index`, this NullPointerException can occur:
```
2021-12-28T22:19:50.056029Z I i.q.g.SqlCompiler plan [q=`select-choose ticker, sample_time, value from (select [ticker, sample_time, value] from symboltest timestamp (sample_time) where ticker = $1)`, fd=1944]
2021-12-28T22:19:50.056840Z E server-main unhandled error [job=io.questdb.cutlass.pgwire.PGWireServer$1@2bedbff1, ex=
java.lang.NullPointerException
        at io.questdb.griffin.FunctionParser.getBindVariableService(FunctionParser.java:873)
        at io.questdb.griffin.FunctionParser.createIndexParameter(FunctionParser.java:158)
        at io.questdb.griffin.FunctionParser.parseIndexedParameter(FunctionParser.java:887)
        at io.questdb.griffin.FunctionParser.createBindVariable(FunctionParser.java:142)
        at io.questdb.griffin.SqlCodeGenerator.generateTableQuery(SqlCodeGenerator.java:2663)
        at io.questdb.griffin.SqlCodeGenerator.generateNoSelect(SqlCodeGenerator.java:1182)
        at io.questdb.griffin.SqlCodeGenerator.generateSelect(SqlCodeGenerator.java:1646)
        at io.questdb.griffin.SqlCodeGenerator.generateQuery0(SqlCodeGenerator.java:1300)
        at io.questdb.griffin.SqlCodeGenerator.generateQuery(SqlCodeGenerator.java:1289)
        at io.questdb.griffin.SqlCodeGenerator.generateSubQuery(SqlCodeGenerator.java:2373)
        at io.questdb.griffin.SqlCodeGenerator.generateSelectChoose(SqlCodeGenerator.java:1859)
        at io.questdb.griffin.SqlCodeGenerator.generateSelect(SqlCodeGenerator.java:1631)
        at io.questdb.griffin.SqlCodeGenerator.generateQuery0(SqlCodeGenerator.java:1300)
        at io.questdb.griffin.SqlCodeGenerator.generateQuery(SqlCodeGenerator.java:1289)
        at io.questdb.griffin.SqlCodeGenerator.generate(SqlCodeGenerator.java:679)
        at io.questdb.griffin.SqlCompiler.generate(SqlCompiler.java:1984)
        at io.questdb.griffin.SqlCompiler.compileUsingModel(SqlCompiler.java:1638)
        at io.questdb.griffin.SqlCompiler.compile(SqlCompiler.java:913)
        at io.questdb.cutlass.pgwire.PGConnectionContext.compileQuery(PGConnectionContext.java:1042)
        at io.questdb.cutlass.pgwire.PGConnectionContext.parseQueryText(PGConnectionContext.java:1435)
        at io.questdb.cutlass.pgwire.PGConnectionContext.processParse(PGConnectionContext.java:1934)
        at io.questdb.cutlass.pgwire.PGConnectionContext.parse(PGConnectionContext.java:1384)
        at io.questdb.cutlass.pgwire.PGConnectionContext.handleClientOperation(PGConnectionContext.java:371)
        at io.questdb.cutlass.pgwire.PGJobContext.handleClientOperation(PGJobContext.java:68)
        at io.questdb.cutlass.pgwire.PGWireServer$1.lambda$$0(PGWireServer.java:72)
        at io.questdb.network.AbstractIODispatcher.processIOQueue(AbstractIODispatcher.java:162)
        at io.questdb.cutlass.pgwire.PGWireServer$1.run(PGWireServer.java:87)
        at io.questdb.mp.Worker.run(Worker.java:111)
]

```

### To reproduce

The issue seems to occur consistently on a 'fresh' (restarted) server instance.

With a table like:
```
CREATE TABLE symboltest (
    ticker symbol index,
    sample_time timestamp,
    value int
) timestamp (sample_time);
INSERT INTO symboltest VALUES ('ABC',NOW(),rnd_int());
```

Then
```
SELECT * FROM symboltest WHERE ticker=?
setString(1,'ABC')
```

If the table is only created without an index on symbol (`CREATE TABLE symboltest ( ticker symbol, ..`) the NPE does not occur.

If a non-prepared statement is executed before hand over JDBC (`SELECT * FROM symboltest WHERE ticker='ABC'`) then sometimes the PreparedStatement does return successfully.

### Expected Behavior

No NullPointerException on the server side, Prepared statement returns successfully.

### Environment

```markdown
- **QuestDB version**: 6.1.3
- **OS**: Windows 10, Zulu JDK 11
- **PostgreSQL driver**: Up to and including 42.3.1
```

### Additional context

Minimal test client:
```
public class TestJDBCSymbol {    
    public static void printResultSet(ResultSet rs) throws SQLException {
        while (rs.next()) {
            String symbol = rs.getString("ticker");
            LocalDateTime timestamp = rs.getObject("sample_time", LocalDateTime.class);
            int value = rs.getInt("value");
            System.out.format("Symbol: %s Timestamp: %s value: %d\n", symbol, timestamp, value);
        }
    }

    public static void main(String[] args) throws SQLException {
        Properties properties = new Properties();
        properties.setProperty("user", "admin");
        properties.setProperty("password", "quest");
        properties.setProperty("sslmode", "disable");

        final Connection connection = DriverManager.getConnection("jdbc:postgresql://localhost:8812/qdb", properties);
        System.out.println("Connected");

        System.out.println("Non prepared statement: ");
        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery("SELECT * FROM symboltest where ticker='ABC'");
        printResultSet(rs);
        
        System.out.println("PreparedStatement:");
        // PreparedStatement version
        PreparedStatement preparedStatement = connection.prepareStatement("SELECT * FROM symboltest where ticker=?");
        preparedStatement.setString(1, "ABC");
        ResultSet preparedRs = preparedStatement.executeQuery();
        printResultSet(preparedRs);

        System.out.flush();
        connection.close();
    }
}
```

### PR DESCRIPTION
**Pull request by bluestreak01**
Closes #1740

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 9 / 9 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|8|8|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 9 / 9 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|8|8|100.00%|

File: core/src/main/java/io/questdb/griffin/FunctionParser.java [hunk 1]
```diff
@@ -132,7 +132,8 @@ public static ScalarFunction createColumn(int position, CharSequence name, Recor
         }
     }
 
-    public Function createBindVariable(int position, CharSequence name) throws SqlException {
+    public Function createBindVariable(SqlExecutionContext sqlExecutionContext, int position, CharSequence name) throws SqlException {
+        this.sqlExecutionContext = sqlExecutionContext;
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/FunctionParser.java:136]: nit: we could also make `FunctionParser` mutable and set `sqlExecutionContext` to null in its `clear()` method. Then `SqlCodeGenerator#clear()` could clear the parser. This way missing initialization will immediately manifest itself through NPEs.

**Review by puzpuzpuz — APPROVED** [review]

File: core/src/main/java/io/questdb/griffin/FunctionParser.java (hunk 1 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/FunctionParser.java:141]: thank you Andrey, makes sense https://github.com/questdb/questdb/pull/1742/commits/2df5cfb39e1d753e934c240203b96cea1e2c8732

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 47 / 47 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryCache.java|11|11|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 47 / 47 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryCache.java|11|11|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 47 / 47 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryCache.java|11|11|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 47 / 47 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryCache.java|11|11|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 47 / 47 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryCache.java|11|11|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 47 / 47 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryCache.java|11|11|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 47 / 47 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|15|15|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactoryCache.java|11|11|100.00%|

**Pull request merged**